components/file_system/src/rate_limiter.rs

```diff
@@ -1,148 +1,568 @@
+    fn new() -> Self {
+        PriorityBasedIORateLimiterProtected {
+            next_refill_time: Instant::now_coarse() + DEFAULT_REFILL_PERIOD,
+            pending_bytes: [0; IOPriority::COUNT],
+            estimated_bytes_through: [IOThroughputEstimator::new(); IOPriority::COUNT],
+        }
+    }
+}
+
+#[derive(Debug, Copy, Clone)]
+struct IOThroughputEstimator {
+    /// total count of sampled epochs
+    count: usize,
+    /// sum of IOs amount over a window
+    sum: usize,
+}
+
+impl IOThroughputEstimator {
+    fn new() -> IOThroughputEstimator {
+        IOThroughputEstimator { count: 0, sum: 0 }
+    }
+
+    fn maybe_update_estimation(&mut self, v: usize) -> Option<usize> {
+        const WINDOW_SIZE: usize = 5;
+        self.count += 1;
+        self.sum += v;
+        // average over DEFAULT_REFILL_PERIOD*WINDOW_SIZE=200ms window
+        if self.count % WINDOW_SIZE == 0 {
+            let avg = self.sum / WINDOW_SIZE;
+            self.sum = 0;
+            Some(avg)
+        } else {
+            None
+        }
+    }
+}
+
+/// Actual implementation for requesting IOs from PriorityBasedIORateLimiter.
+/// An attempt will be recorded first. If the attempted amount exceeds the available quotas of
+/// current epoch, the requester will register itself and sleep until next epoch.
+macro_rules! request_imp {
+    ($self:ident, $priority:ident, $amount:ident, $mode:tt) => {{
+        let priority_idx = $priority as usize;
+        loop {
+            let cached_bytes_per_refill =
+                $self.bytes_per_epoch[priority_idx].load(Ordering::Relaxed);
+            if cached_bytes_per_refill == 0 {
+                return $amount;
+            }
+            let amount = std::cmp::min($amount, cached_bytes_per_refill);
+            let bytes_through =
+                $self.bytes_through[priority_idx].fetch_add(amount, Ordering::AcqRel) + amount;
+            if bytes_through <= cached_bytes_per_refill {
+                return amount;
+            }
+            let now = Instant::now_coarse();
+            let (next_refill_time, pending) = {
+                let mut locked = $self.protected.lock();
+                // a small delay in case a refill slips in after `bytes_per_epoch` was fetched.
+                if locked.next_refill_time + Duration::from_millis(1) >= now + DEFAULT_REFILL_PERIOD
+                {
+                    continue;
+                }
+                locked.pending_bytes[priority_idx] += amount;
+                (locked.next_refill_time, locked.pending_bytes[priority_idx])
+            };
+            if next_refill_time > now {
+                let wait = next_refill_time - now;
+                RATE_LIMITER_REQUEST_WAIT_DURATION
+                    .with_label_values(&[$priority.as_str()])
+                    .observe(wait.as_secs_f64());
+                do_sleep!(wait, $mode);
+            } else if next_refill_time + DEFAULT_REFILL_PERIOD / 2 < now {
```

Comment:
let's see if we can get rid of background task, and use the request thread to do the refill.